apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/ApmServerConfigurationSource.java

```diff
@@ -0,0 +1,108 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2019 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.configuration;
+
+import co.elastic.apm.agent.impl.MetaData;
+import co.elastic.apm.agent.report.ApmServerClient;
+import co.elastic.apm.agent.report.serialize.PayloadSerializer;
+import com.dslplatform.json.DslJson;
+import com.dslplatform.json.JsonReader;
+import com.dslplatform.json.MapConverter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.stagemonitor.configuration.source.AbstractConfigurationSource;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.util.Collections;
+import java.util.Map;
+
+public class ApmServerConfigurationSource extends AbstractConfigurationSource {
+    private static final Logger logger = LoggerFactory.getLogger(ApmServerConfigurationSource.class);
+
+    private final DslJson<Object> dslJson = new DslJson<>();
+    private final byte[] buffer = new byte[4096];
+    private final PayloadSerializer payloadSerializer;
+    private final MetaData metaData;
+    private final ApmServerClient apmServerClient;
+    @Nullable
+    private String etag;
+    private Map<String, String> config = Collections.emptyMap();
```

Comment:
```suggestion
    private volatile Map<String, String> config = Collections.emptyMap();
```

Reply:
Not needed, as the configuration is only ready from the same thread. The values are then updated on the `ConfigurationOption` class which only parses the String if it differs from the previous value.